Refactor host files into crates

lukewilliamboswell opened the pull request:
This PR separates the rust host files from the roc platform files, and adds a `build.roc` script.

The purpose of this change is to simplify the build process, and to have a clearer separation between these elements as this repository serves as an example of roc platforms.

smores56 commented on README.md:
````diff
@@ -14,18 +14,20 @@ Behind the scenes, `basic-webserver` uses Rust's high-performance [hyper](https:
+```roc
+app [main] {
+    # replaced the pf: "../platform/main.roc" with the following
+     pf: "https://github.com/roc-lang/basic-webserver/releases/<latest release hash>.tar.br"
````
This should be `pf: platform "..."`, right?

lukewilliamboswell replied:
Done

smores56 reviewed (approved):
Looks good! We may want to wait on merging for @Anton-4 to finish the https://github.com/roc-lang/roc/issues/6873 investigation that is blocking the sister PR for basic-cli here: https://github.com/roc-lang/basic-cli/pull/194

lukewilliamboswell commented on build.roc:
```diff
@@ -0,0 +1,118 @@
+app [main] {
+    cli: platform "https://github.com/roc-lang/basic-cli/releases/download/0.12.0/cf_TpThUd4e69C7WzHxCbgsagnDmk3xlb_HmEKXTICw.tar.br",
+    weaver: "https://github.com/smores56/weaver/releases/download/0.2.0/BBDPvzgGrYp-AhIDw0qmwxT0pWZIQP_7KOrUrZfp_xw.tar.br",
```
Done

lukewilliamboswell commented:
- Updated with changes to support the [List refcounting performance fix](https://github.com/roc-lang/roc/pull/6894). 
- Refactored Http to align closer to basic-cli and remove the glue generated types. Added `roc_http` and refactored the `roc_host` crate.
- Fixed up and simplified the glue generation. Moved the `glue-get.sh` script to top level so it is easier to discover. The generated code still requires some manual clean-up but this is much simpler now. 
- Also includes a manual `impl<T, E> RocRefcounted for RocResult<T, E>` in `roc_std` that needs to be added into `roc-lang/roc/crates/roc_std`

Anton-4 commented on .github/workflows/ci.yml:
Let's do formatting changes in a separate PR in the future.

Anton-4 commented on crates/roc_app/src/wasm32.rs:
```diff
@@ -1429,140 +1550,236 @@ impl PartialOrd for ConnectResult {
 }
 
+impl roc_std::RocRefcounted for StreamErr {
+    fn inc(&mut self) {
+        unimplemented!();
+    }
+    fn dec(&mut self) {
+        unimplemented!();
+    }
+    fn is_refcounted() -> bool {
+        true
+    }
```
This seems problematic?

bhansconnect replied:
Only matters if a list containing said type is passed between host and roc.

So if there are no effects of the sort `List (Result Stream StremErr)`, doesn't matter.

Same witht he other case pointed out.

Anton-4 commented on crates/roc_app/src/wasm32.rs:
```diff
@@ -1654,124 +1871,165 @@ impl PartialOrd for WriteResult {
     }
 }
 
+impl roc_std::RocRefcounted for ReadResult {
+    fn inc(&mut self) {
+        unimplemented!();
+    }
+    fn dec(&mut self) {
+        unimplemented!();
+    }
+    fn is_refcounted() -> bool {
+        true
+    }
+}
```
Same here

lukewilliamboswell replied:
This was also generated by `RustGlue.roc`. I think we could remove it if we wanted to, because it's not used or needed. I think the implementation is just to satisfy the trait requirement.

Anton-4 replied:
Thanks, I'll try to remove them all

Anton-4 commented:
Looks like some of these are used...

bhansconnect commented:
> Looks like some of these are used...

Oh, I know whats happening. They type is being used in a RocResult. Which does use the refcounting impl. But still, the actual impl won't be called unless eventually the type is used in a list.

bhansconnect commented:
So they are unused, but referenced by RocResult. So they can't be deleted.

Anton-4 commented on crates/README.md:
```diff
@@ -0,0 +1,18 @@
+TODO other crates
```
I just want to have this handled before we merge. I can probably get to it tomorrow but anyone is welcome to take it.

Anton-4 replied:
> I can probably get to it tomorrow

I got deep down a segfault rabbit hole when updating the examples to basic-cli 0.13 :p

Anton-4 replied:
@lukewilliamboswell  Can you explain what the purpose is of the `roc_app` and `roc_fn` crates and why we don't need those in basic-cli?

Anton-4 reviewed (approved):
Thank you very much @lukewilliamboswell!
